pkg/apis/kubeflow.org/v2alpha1/trainingruntime_types.go

```diff
@@ -15,3 +15,181 @@ limitations under the License.
+type GangSchedulerPlugin string
+
+const (
+	// Volcano plugin for gang-scheduling.
+	GangSchedulerPluginVolcano GangSchedulerPlugin = "volcano"
+
+	// Coscheduling plugin from the Kubernetes scheduler-plugins for gang-scheduling.
+	GangSchedulerPluginCoscheduling GangSchedulerPlugin = "coscheduling"
+)
+
+// MLSpec represents the runtime-specific configuration for various technologies.
+// One of the following specs can be set.
+type MLSpec struct {
+	// Configuration for the PyTorch runtime.
+	TorchSpec *TorchSpec `json:"torchSpec,omitempty"`
+
+	// Configuration for the MPI Runtime.
+	MPISpec *MPISpec `json:"mpiSpec,omitempty"`
+}
+
+// TorchSpec represents a PyTorch runtime configuration.
+type TorchSpec struct {
+	// Number of processes per node.
+	// This value is inserted into the `--nproc-per-node` argument of the `torchrun` CLI.
+	// Supported values: `auto`, `cpu`, `gpu`, or int value.
+	// TODO (andreyvelich): Add kubebuilder validation.
+	// Defaults to `auto`.
+	NumProcPerNode *string `json:"numProcPerNode,omitempty"`
+
+	// Whether to run single-node multi-worker training.
+	// This value is inserted into the `--standalone` argument of the `torchrun` CLI.
+	// Defaults to false.
+	Standalone *bool `json:"standalone,omitempty"`
+
+	// Elastic policy for the PyTorch training.
+	ElasticPolicy *TorchElasticPolicy `json:"elasticPolicy,omitempty"`
+}
+
+// TorchElasticPolicy represents a configuration for the PyTorch elastic training.
+// If this policy is set, the `.spec.numNodes` parameter must be omitted, since min and max node
+// is used to configure the `torchrun` CLI argument: `--nnodes=minNodes:maxNodes`.
+// Only `c10d` backend is supported for the Rendezvous communication.
+type TorchElasticPolicy struct {
+	// How many times the training job can be restarted.
+	// This value is inserted into the `--max-restarts` argument of the `torchrun` CLI and
+	// the `.spec.failurePolicy.maxRestarts` parameter of the training Job.
+	MaxRestarts *int32 `json:"maxRestarts,omitempty"`
+
+	// Lower limit for the number of nodes to which training job can scale down.
+	MinNodes *int32 `json:"minNodes,omitempty"`
+
+	// Upper limit for the number of nodes to which training job can scale up.
+	MaxNodes *int32 `json:"maxNodes,omitempty"`
+
+	// Specification which are used to calculate the desired number of nodes. See the individual
+	// metric source types for more information about how each type of metric must respond.
+	// The HPA will be created to perform auto-scaling.
+	Metrics []autoscalingv2.MetricSpec `json:"metrics,omitempty"`
+}
+
+// MPISpec represents a MPI runtime configuration.
+type MPISpec struct {
+	// Number of processes per node.
+	// This value is equal to the number of slots for each node in the hostfile.
+	NumProcPerNode *int32 `json:"numProcPerNode,omitempty"`
+
+	// Implementation name for the MPI to create the appropriate hostfile.
+	MPIImplementation *MPIImplementation `json:"mpiImplementation"`
```

Comment:
Since MPIImplementation is a const, does this need to be a pointer? With a pointer, a `nil` is a valid value. Do we want user to pass a nil value? 

Also with `omitempty` not specified, the zero value of this would be `nil`. The zero value will be included in all patch requests which seems unnecessary.

Maybe you meant to include the `omitempty` but accept a nil value?

Reply:
Yes, I think we should remove pointer from here. @tenzen-y What do you think ?

Reply:
```go
	// Implementation name for the MPI to create the appropriate hostfile.
	// Defaults to OpenMPI
	MPIImplementation *MPIImplementation `json:"mpiImplementation,omitempty"`
```

In most cases of optional fields, we should use the `omitempty` tag.
Please refer to more details here: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required